app/scripts/lib/extension-port-stream.ts

```diff
@@ -0,0 +1,398 @@
+/**
+ * Creates a random ID for chunked messages. Always <= Int32 max value.
+ *
+ * @returns a random number that can be used as an ID
+ */
+function getNextId(): number {
+  // increment the counter and wrap around if it exceeds Int32 max value
+  _randomCounter = (_randomCounter + 1) & 0x7fffffff;
+  return _randomCounter;
+}
+
+type QueuedEntry = {
+  parts: string[];
+  total: number;
+  received: number;
+};
+
+type Id = number;
+type Seq = number;
+type Total = number;
+type Data = string;
+
+export type ChunkFrame = {
+  id: Id;
+  total: Total;
+  seq: Seq;
+  data: Data;
+};
+
+/**
+ * A transport frame that represents a chunked message.
+ *
+ * @property id - a unique identifier for the chunked message
+ * @property total - the total number of chunks in the message
+ * @property seq - the sequence number of this chunk
+ * @property data - the actual data of this chunk
+ */
+export type TransportChunkFrame = `${Id}|${Total}|${Seq}|${Data}`;
+
+/**
+ * Options for the PortStream.
+ *
+ * @property log - a function to log messages
+ * @property debug - whether to enable debug mode
+ * @property chunkSize - the size of each chunk in bytes
+ */
+export type Options = {
+  /**
+   * A function to log messages.
+   *
+   * @param data - the data to log
+   * @param outgoing - whether the data is outgoing (true) or incoming (false)
+   */
+  log?: (data: unknown, outgoing: boolean) => void;
+
+  /**
+   * Whether to enable debug mode. In debug mode, the stream will log
+   * incoming and outgoing messages to the console.
+   * This is useful for debugging, but should be disabled in production.
+   * Defaults to `false`.
+   */
+  debug?: boolean;
+
+  /**
+   * The size of each chunk in bytes. Set to 0 to disabled chunking.
+   */
+  chunkSize?: number;
+} & DuplexOptions;
+
+/**
+ * The default chunk size for the stream.
+ *
+ * Chromium has a maximum message size of 64 MB.
+ *
+ * FireFox's doesn't really have a maximum limit (I've tested it up to 2 GB), but
+ * it does limit the size of the strings you can send to 1 GB per string.
+ *
+ * So we can just send messages on FireFox without chunking.
+ *
+ * Chrome limit: 64 * 1024 * 1024 - 1 (64 MB - 1 byte)
+ * Firefox limit: 0 (no chunking, send as is)
+ */
+export const CHUNK_SIZE = globalThis.navigator.userAgent
+  .toLowerCase()
+  .includes('firefox')
+  ? 0
+  : 64 * 1024 * 1024 - 1;
+
+enum S {
+  Id,
+  Total,
+  Seq,
+  Data,
+}
+
+/**
+ * Parses a chunk frame from a string.
+ *
+ * a chunk frame is in the form: `<id>|<total>|<seq>|<data>`, where:
+ * - `<id>` is a unique number identifier for the chunked message
+ * - `<total>` is the total number of chunks in the message
+ * - `<seq>` is the sequence number of this chunk (0-based)
+ * - `<data>` is the actual data of this chunk
+ *
+ * @param input - the value to parse
+ * @returns a ChunkFrame if the value is a valid chunk frame, otherwise false
+ */
+const maybeParseChunkFrame = (input: unknown): ChunkFrame | false => {
```

Comment:
The state-machine parser below is about 10x faster than string manipulating parsing (indexOf/split/parseInt) and over 100x faster then regular expression based parsing.